### ISSUE TITLE
Event handlers not being updated on prop changes

### ISSUE THREAD

**Issue by redbmk** [alex3165/react-mapbox-gl#442]
Event handlers appear to only be [added when the map is created](https://github.com/alex3165/react-mapbox-gl/blob/61aa3aeb94219526eb24481e2ab2b9ebdd8379b1/src/map.tsx#L322-L330), in `componentDidMount`. When the property changes to either add or remove them from the `react-mapbox-gl` component, those changes are not reflected in the map.

Here's a simple example showcasing two scenarios where you can see the issue:

```
state = {
  trackMovements: false,
  loadedData: false,
};

// this should only get called once, but instead gets called repeatedly
// as the source data is loaded (e.g. while the map is zoomed in or moved around)
getData = (map, { sourceId }) => {
  this.setState({ loadedData: true, sourceId });
};

// once the button is clicked, this should get called every time the map moves,
// but instead will never get called.
onMove = (map) => {
  console.log(map.getCenter());
};

render() {
  <button onClick={() => this.setState({ trackMovements: true })}>Track movements</button>
  <Map
    onMove={this.state.trackMovements && this.onMove}
    onSourceData={!this.state.loadedData && this.getData}
  />
}
```

**Comment by alex3165** [comment]
Indeed, this need to be added, although as a work around for now you could do:
```js
state = {
  trackMovements: false,
  loadedData: false,
};

// this should only get called once, but instead gets called repeatedly
// as the source data is loaded (e.g. while the map is zoomed in or moved around)
getData = (map, { sourceId }) => {
  if (this.state.loadedData) { return; }
  this.setState({ loadedData: true, sourceId });
};

// once the button is clicked, this should get called every time the map moves,
// but instead will never get called.
onMove = (map) => {
  if (!this.state.trackMovements) { return; }
  console.log(map.getCenter());
};

render() {
  <button onClick={() => this.setState({ trackMovements: true })}>Track movements</button>
  <Map
    onMove={this.onMove}
    onSourceData={this.getData}
  />
}
```

**Comment by alex3165** [comment]
https://github.com/alex3165/react-mapbox-gl/pull/446/files should fix it

**Comment by redbmk** [comment]
Woah, nice - I didn't expect a fix to that so quick!

**Comment by alex3165** [comment]
Released from tag `3.0.0`
